### ISSUE TITLE
Casting to a looser type results in an error

### ISSUE THREAD

**Issue by MarkKahn** [facebook/flow#6553]
Simplified version of what I'm running in to:

    (({a: 42}: {a?: number}): {a?: number, b?: number})

Flow complains because the first object doesn't contain `b`, but `b` is optional...

**Comment by jcready** [comment]
That's because the first object type `{a?: number}` is inexact which means that it can have properties other than `a`. For instance this passes type checking: `({a: 42, b: 'string'}: {a?: number})`. Therefore it should be obvious why you cannot cast `{a?: number}` to `{a?: number, b?: number}` since any object value occupying `{a?: number}` could have any value for `b` not just `number | void`.

**Comment by MarkKahn** [comment]
@jcready — the same issue occurs if you make the first type exact though, as in `{|a?: number|}`

**Comment by jcready** [comment]
If you mark `b` as covariant then it will work:

```js
(({a: 42}: {|a?: number|}): {a?: number, +b?: number})
```

**Comment by MarkKahn** [comment]
@jcready -- Then `b` is unwritable not just for this but for anything using the type.  Is there anything logically wrong with:

    (({a: 42}: {|a?: number|}): {a?: number, b?: number})

?  If not, this is a bug in flow, no?

Perhaps a use case would help:

**Comment by wchargin** [comment]
> Is there anything logically wrong with:
> 
> ```js
> (({a: 42}: {|a?: number|}): {a?: number, b?: number})
> ```

Yes: `{|a?: number|}` is not a subtype of `{a?: number, b?: number}`.
And it cannot be: if it were, then we could write the following:

```js
function addB(x: {a?: number, b?: number}) {
  // No problem: `b` is an optional, writable property.
  x.b = 3;
}
function assertNoB(x: {|a?: number|}) {
  // In a typesafe program, this assertion should always pass, because
  // the type of `x` is exact and does not include `b`.
  if ("b" in x) {
    throw new Error("!");
  }
}

const x: {|a?: number|} = {a: 1};
const y: {a?: number, b?: number} = x;  // if this is legal...
addB(y);  // ...then this is legal...
assertNoB(x);  // but this fails!
```

This is why marking `b` as read-only solves the problem.
